File: strategy/generator/signal/zerocross/tsi.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    StaticParameter,
)
from core.models.smooth import Smooth
from core.models.source import SourceType
from strategy.generator.signal.base import Signal, SignalType


@dataclass(frozen=True)
class TsiZeroCrossSignal(Signal):
    type: SignalType = SignalType.TsiZeroCross
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    smooth_type: Parameter = StaticParameter(Smooth.EMA)
    fast_period: Parameter = StaticParameter(13.0)
    slow_period: Parameter = StaticParameter(25.0)

File: ta_lib/strategies/signal/src/zerocross/tsi.rs
use base::prelude::*;
use core::prelude::*;
use momentum::tsi;
use timeseries::prelude::*;

pub struct TsiZeroCrossSignal {
    source_type: SourceType,
    smooth_type: Smooth,
    fast_period: usize,
    slow_period: usize,
}

impl TsiZeroCrossSignal {
    pub fn new(
        source_type: SourceType,
        smooth_type: Smooth,
        fast_period: f32,
        slow_period: f32,
    ) -> Self {
        Self {
            source_type,
            smooth_type,
            fast_period: fast_period as usize,
            slow_period: slow_period as usize,
        }
    }
}

impl Signal for TsiZeroCrossSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.fast_period, self.slow_period)
    }

    fn trigger(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let tsi = tsi(
            &data.source(self.source_type),
            self.smooth_type,
            self.slow_period,
            self.fast_period,
        );

        (tsi.cross_over(&ZERO), tsi.cross_under(&ZERO))
    }
}
